File: firefly/src/main/java/com/firefly/server/http2/router/impl/AbstractPreciseMatcher.java
package com.firefly.server.http2.router.impl;

import com.firefly.server.http2.router.Matcher;
import com.firefly.server.http2.router.Router;

import java.util.*;

/**
 * @author Pengtao Qiu
 */
abstract public class AbstractPreciseMatcher implements Matcher {

    protected Map<String, Set<Router>> map;

    private Map<String, Set<Router>> map() {
        if (map == null) {
            map = new HashMap<>();
        }
        return map;
    }

    @Override
    public void add(String rule, Router router) {
        map().computeIfAbsent(rule, k -> new HashSet<>()).add(router);
    }

    @Override
    public MatchResult match(String value) {
        if (map == null) {
            return null;
        }

        Set<Router> routers = map.get(value);
        if (routers != null && !routers.isEmpty()) {
            return new MatchResult(routers, Collections.emptyMap(), getMatchType());
        } else {
            return null;
        }
    }

}

File: firefly-net/src/main/kotlin/com/fireflysource/net/http/server/impl/matcher/AbstractPreciseMatcher.kt
package com.fireflysource.net.http.server.impl.matcher

import com.fireflysource.net.http.server.Matcher
import com.fireflysource.net.http.server.Router
import java.util.*

abstract class AbstractPreciseMatcher : AbstractMatcher<String>(), Matcher {

    override fun add(rule: String, router: Router) {
        routersMap.computeIfAbsent(rule) { TreeSet() }.add(router)
    }

    override fun match(value: String): Matcher.MatchResult? {
        val routers = routersMap[value]
        return if (!routers.isNullOrEmpty()) {
            Matcher.MatchResult(routers, emptyMap(), matchType)
        } else null
    }
}